### PR TITLE
Feat: replace plugin RPC with BTP

### PR DESCRIPTION
**Pull request by sharafian**

### CODE REVIEW

**Comment by sharafian** [comment]
Should be ready to merge after review

**Comment by sharafian** [comment]
sorry, accidentally hit close and comment

File: test/infoSpec.js [hunk 1]
```diff
@@ -129,6 +130,133 @@ describe('Info', () => {
...
+      await this.plugin.addSocket(this.newSocket)
+    })
+
+    it('should deny an authentication request with wrong method', async function () {
+      this.newSocket.emit('message', btpPacket.serializeFulfill({
+        transferId: 'b38a5203-bdb8-f11f-db01-5a32cf1a4e43',
+        fulfillment: 'Ndr_HMuLPPl0idUlvAXFXBVQTFOizq-nXozej0KIA7k'
+      }, 100, []))
+
+      await new Promise((resolve) => {
```
Comment (dappelt) [inline test/infoSpec.js]: You don't need to wrap this in a promise. `this.mockSocket.isDone()` makes sure all handlers registered via `.reply()` have been called.

File: test/infoSpec.js [hunk 2]
```diff
@@ -129,6 +130,133 @@ describe('Info', () => {
     })
   })
 
+  describe('authentication', () => {
+    beforeEach(async function () {
+      this.newSocket = new MockSocket()
+      await this.plugin.addSocket(this.newSocket)
+    })
+
```
Comment (dappelt) [inline test/infoSpec.js]: Check that all `MockSocket` handlers have been called by adding `afterEach`:

```js
  afterEach(async function () {
    assert(await this.mockSocket.isDone(), 'request handlers must have been called')
  })
```

File: test/infoSpec.js [hunk 3]
```diff
@@ -129,6 +130,133 @@ describe('Info', () => {
...
+          assert.equal(e.data.data, '{"message":"invalid method on unauthenticated socket"}')
+          resolve()
+        })
+      })
+    })
+
+    it('should deny an authentication request with no "auth" protocol', async function () {
+      this.newSocket.emit('message', btpPacket.serializeMessage(100, []))
+
+      await new Promise((resolve) => {
```
Comment (dappelt) [inline test/infoSpec.js]: No need for this Promise

File: test/infoSpec.js [hunk 4]
```diff
@@ -129,6 +130,133 @@ describe('Info', () => {
...
+    })
+
+    it('should deny an authentication request with no "auth_token" protocol', async function () {
+      this.newSocket.emit('message', btpPacket.serializeMessage(100, [{
+        protocolName: 'auth',
+        contentType: btpPacket.MIME_APPLICATION_OCTET_STREAM,
+        data: Buffer.from('')
+      }]))
+
+      await new Promise((resolve) => {
```
Comment (dappelt) [inline test/infoSpec.js]: No need for this Promise

File: test/infoSpec.js [hunk 5]
```diff
@@ -129,6 +130,133 @@ describe('Info', () => {
...
+        protocolName: 'auth',
+        contentType: btpPacket.MIME_APPLICATION_OCTET_STREAM,
+        data: Buffer.from('')
+      }, {
+        protocolName: 'auth_token',
+        contentType: btpPacket.MIME_TEXT_PLAIN_UTF8,
+        data: Buffer.from('')
+      }]))
+
+      await new Promise((resolve) => {
```
Comment (dappelt) [inline test/infoSpec.js]: No need for this Promise

File: test/infoSpec.js [hunk 6]
```diff
@@ -129,6 +130,133 @@ describe('Info', () => {
...
+        protocolName: 'auth_token',
+        contentType: btpPacket.MIME_TEXT_PLAIN_UTF8,
+        data: Buffer.from('garbage')
+      }, {
+        protocolName: 'auth_username',
+        contentType: btpPacket.MIME_TEXT_PLAIN_UTF8,
+        data: Buffer.from('')
+      }]))
+
+      await new Promise((resolve) => {
```
Comment (dappelt) [inline test/infoSpec.js]: No need for this Promise

File: test/infoSpec.js [hunk 7]
```diff
@@ -129,6 +130,133 @@ describe('Info', () => {
...
+        protocolName: 'auth_token',
+        contentType: btpPacket.MIME_TEXT_PLAIN_UTF8,
+        data: Buffer.from('placeholder')
+      }, {
+        protocolName: 'auth_username',
+        contentType: btpPacket.MIME_TEXT_PLAIN_UTF8,
+        data: Buffer.from('')
+      }]))
+
+      await new Promise((resolve) => {
```
Comment (dappelt) [inline test/infoSpec.js]: No need for this Promise

**Review by dappelt — CHANGES_REQUESTED** [review]

**Comment by michielbdejong** [comment]
It's a bit ugly that you need to specify both `opts.prefix` and `opts.info.prefix`, can you copy the second from the first?

**Review by michielbdejong — APPROVED** [review]

File: test/infoSpec.js (hunk 2 above)
Comment (sharafian) [reply test/infoSpec.js]: added

File: src/lib/client.js [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+    // See also: https://github.com/interledger/rfcs/pull/300
+    const parsedBtpUri = new URL(server)
+    this._authUsername = parsedBtpUri.username
+    this._authToken = parsedBtpUri.password
+    parsedBtpUri.username = ''
+    parsedBtpUri.password = ''
+    // Note that setting the parsedBtpUri.protocol does not work as expected,
+    // so removing the 'btp+' prefix from the full URL here:
+    assert(parsedBtpUri.toString().startsWith('btp+'), 'server uri must start with "btp+"')
+    this._wsUri = parsedBtpUri.toString().substring('btp+'.length)
```
Comment (emschwartz) [inline src/lib/client.js]: This seems like a bit of a janky way to either input `ws` or `wss`. It should be more explicit and throw an error if it's anything else

Comment (emschwartz) [reply src/lib/client.js]: Also shouldn't it take out the auth details?

File: test/channelSpec.js [hunk 9]
```diff
@@ -52,8 +57,9 @@ describe('makePaymentChannelPlugin', function () {
     this.PluginClass = makePaymentChannelPlugin(this.channel)
     this.plugin = new (this.PluginClass)(this.opts)
 
-    this.fulfillment = require('crypto').randomBytes(32)
-    this.transfer = {
+    this.fulfillment = crypto.randomBytes(32)
```
Comment (emschwartz) [inline test/channelSpec.js]: Probably better to just use a fixed value here

**Review by emschwartz — COMMENTED** [review]
Overall looks okay. This module has a lot going on in it and it can be used in a lot of different ways so it might be worth breaking it up more

File: src/lib/client.js (hunk 8 above)
Comment (dappelt) [reply src/lib/client.js]: Yes, it should take out auth details. That's done [here](https://github.com/interledgerjs/ilp-plugin-payment-channel-framework/pull/21/files/c14320143b174a9c4262d320c0d804eecaa8fe5a#diff-976824b6113e95afacd333cbfab6d019R22).

**Review by dappelt — APPROVED** [review]

**Comment by sharafian** [comment]
Merging without tests passing because integration tests time out. This repo isn't used in the integration tests so it isn't an issue with the PR.

**Pull request merged**